### DISCUSSION TITLE
Linux and Links 2003

### DISCUSSION THREAD

**Discussion by dwg1957** [elishacloud/dxwrapper#316] [General]
I am trying to get Links 2003, which is a Windows golf game working under Linux. I have it installed and it does run under Wine but I get course crashes. I have has some success with another wrapper that does resolve the crashed but it has problems with some GPU types (Intel) and some performance problems with Nvidia on  a machine, so I thought I would give dxwrapper a try.

It's a promising sign in that I get the game running with dxwrapper installed a Dd7to9 set. However On initial starting I am seeing a couple of problems.

The first is that the golfer disappears, or parts of him does, the ball is disappearing at times and what should be reflective water i.e. water in say a lake that reflects tree, sky etc it just showing as light blue.

I have absolutely no idea what to look at or what to try changing, I've attached the log file if that is of any help

[dxwrapper-linksmmiii.log](https://github.com/user-attachments/files/17676200/dxwrapper-linksmmiii.log)

**Comment by elishacloud** [comment]
I noticed a lot of these entries in your log file:

```log
392 20:20:44.924 m_IDirectDrawSurfaceX::CopySurface Error: could not copy rect: {{ DDSCAPS_LOCALVIDMEM  DDSCAPS_VIDEOMEMORY  DDSCAPS_ZBUFFER },0x0,0x0,0x0} -> {{ DDSCAPS_SYSTEMMEMORY  DDSCAPS_ZBUFFER },0x0,0x0,0x0} D3DFMT_D24S8 -> D3DFMT_D24S8 {813,290,1045,301} -> {0,0,232,11} 0 D3DERR_INVALIDCALL
392 20:20:44.924 D3DXLoadSurfaceFromSurface Error: Failed to Copy surface!
```

I suspect that is the issue.  Copying depth buffer surfaces sub-rectangles are not supported in Direct3D9.  I might have to created a shader or find another way to do that.

However, one thing you could try is using the Windows XP build of dxwrapper.  This removes some of the hooks that are not needed under Linux.

**Comment by dwg1957** [comment]
Thanks for the response.

Downloaded winxp.binaries.zip replaced the existing files in the games folder. Alas the same problem.

**Comment by dwg1957** [comment]
FWIW on Linux I have found using Wine alone with Links works the best. It took some wrangling around what looks like some system issues as the same problems were found under Windows on the same machine.

I have decided to see if I could get dxwrapper working with Links on Windows on another system as I am having a few problems with some courses when using the recommended dgVoodoo.

I get the best result by using the setting EnableDdrawWrapper = 1 and set nothing else.

The games main menu comes up, I can load a golf course and it comes up but popup menus are not displayed correctly and my golfer disappears

If I also set Dd7to9 = 1 the game will not load

I'm at a loss of what settings to try, the ones I have either do not help or make it worse.

Any advice appreciated. I attach the log from when EnableDdrawWrapper is set to 1

[dxwrapper-linksmmiii.log](https://github.com/user-attachments/files/20966406/dxwrapper-linksmmiii.log)

**Comment by dwg1957** [comment]
I just tested the latest release 1_4_7900_25 On Windows 11 with Links 2003.

I'm seeing the problem I reported earlier with the disappearing golfer and corrupt menus and assuming the changes made in this release are not expected to have any effect on these problems.

Is my assumption correct or am I missing something?

Thanks

Des

**Comment by elishacloud** [comment]
I think the issue with this game is with a video memory zbuffer copy.  dxwrapper doesn't support that yet.  I added a fix for that but it caused issues with some games so I disabled it.

Here is a build with the fix re-enabled.  Let me know if this fixes the issue:  [dxwrapper.zip](https://github.com/user-attachments/files/22029518/dxwrapper.zip)

**Comment by dwg1957** [comment]
[dxwrapper-linksmmiii.log](https://github.com/user-attachments/files/22038524/dxwrapper-linksmmiii.log)

I have tried the new build but alas there has been no change, still the disappearing golfer and corrupted menus when starting a round of golf.

**Comment by elishacloud** [comment]
Ok, I will need to test this.  I am guessing it is a zbuffer issue.  There are still some zbuffer issues I need to fix because d3d9 has many zbuffer limitations.

**Comment by dwg1957** [comment]
I'm not sure if it is relevant but I believe Links is a DirectX 7 application.

There is a trial version of Links, it only includes 9 holes of a fictional course and does not have all options so is not suitable to test everything, but I would think it would show the basic problem.

https://www.linkscountryclub.com/links-2003-trial-version
